File: daemon/provider_check.py
from __future__ import annotations

import json
import os
import shutil
import sys

from daemon.llm import create_provider


def _emit(payload: dict) -> None:
    sys.stdout.write(json.dumps(payload, ensure_ascii=False))


def main() -> int:
    try:
        raw = sys.stdin.read()
        data = json.loads(raw or "{}")
    except json.JSONDecodeError:
        _emit({"ok": False, "code": "invalid_payload"})
        return 1

    provider = str(data.get("provider") or "").strip()
    claude_model = str(data.get("claude_model") or "haiku")
    codex_model = str(data.get("codex_model") or "gpt-5.4")
    gemini_model = str(data.get("gemini_model") or "gemini-3.1-flash-lite-preview")
    gemini_api_key = str(data.get("gemini_api_key") or "").strip()

    if provider == "external":
        _emit({"ok": True, "code": "external"})
        return 0

    if provider not in {"gemini", "claude", "codex"}:
        _emit({"ok": False, "code": "invalid_provider"})
        return 1

    if provider == "gemini":
        api_key = gemini_api_key or os.environ.get("GEMINI_API_KEY", "")
        if not api_key:
            _emit({"ok": False, "code": "missing_api_key"})
            return 0
        os.environ["GEMINI_API_KEY"] = api_key

    if provider == "claude" and not shutil.which("claude"):
        _emit({"ok": False, "code": "missing_binary"})
        return 0

    if provider == "codex" and not shutil.which("codex"):
        _emit({"ok": False, "code": "missing_binary"})
        return 0

    token = f"VIDA_PROVIDER_CHECK_OK_{provider.upper()}"
    prompt = f"Return exactly the string: {token}"

    result = create_provider(
        provider,
        claude_model=claude_model,
        codex_model=codex_model,
        gemini_model=gemini_model,
    ).generate_text(prompt, timeout=45)

    if result and token in result:
        _emit({"ok": True, "code": "ready"})
        return 0

    detail = (result or "").strip()[:200]
    if provider in {"claude", "codex"}:
        _emit({"ok": False, "code": "binary_found_but_failed", "detail": detail})
        return 0

    _emit({"ok": False, "code": "request_failed", "detail": detail})
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: web/src-tauri/src/commands/provider_check.rs
use crate::db::AppDb;
use crate::python;
use serde::Deserialize;
use std::io::Write;
use std::process::{Command, Stdio};
use tauri::State;

#[derive(Deserialize)]
pub struct ProviderCheckRequest {
    provider: String,
    gemini_model: Option<String>,
    claude_model: Option<String>,
    codex_model: Option<String>,
    gemini_api_key: Option<String>,
}

#[tauri::command]
pub async fn validate_provider(
    body: ProviderCheckRequest,
    db: State<'_, AppDb>,
) -> Result<serde_json::Value, String> {
    let provider = body.provider.trim();
    if !matches!(provider, "gemini" | "claude" | "codex" | "external") {
        return Ok(serde_json::json!({
            "ok": false,
            "code": "invalid_provider",
        }));
    }

    if provider == "external" {
        return Ok(serde_json::json!({
            "ok": true,
            "code": "external",
        }));
    }

    let repo_root = &db.config_dir;
    let python = match python::find_python(repo_root) {
        Ok(p) => p,
        Err(_) => {
            return Ok(serde_json::json!({
                "ok": false,
                "code": "python_not_found",
            }));
        }
    };

    let daemon_src = &db.daemon_src;
    let canon_daemon = match std::fs::canonicalize(daemon_src) {
        Ok(p) => p,
        Err(_) => {
            return Ok(serde_json::json!({
                "ok": false,
                "code": "daemon_not_found",
            }));
        }
    };
    let provider_check = canon_daemon.join("daemon").join("provider_check.py");
    if !provider_check.is_file() {
        return Ok(serde_json::json!({
            "ok": false,
            "code": "checker_not_found",
        }));
    }

    let gemini_api_key = body
        .gemini_api_key
        .filter(|v| !v.trim().is_empty())
        .or_else(|| db.get_setting("env.GEMINI_API_KEY"));

    let payload = serde_json::json!({
        "provider": provider,
        "gemini_model": body.gemini_model.unwrap_or_else(|| "gemini-3.1-flash-lite-preview".to_string()),
        "claude_model": body.claude_model.unwrap_or_else(|| "haiku".to_string()),
        "codex_model": body.codex_model.unwrap_or_else(|| "gpt-5.4".to_string()),
        "gemini_api_key": gemini_api_key.unwrap_or_default(),
    });

    let mut cmd = Command::new(&python);
    cmd.arg("-m")
        .arg("daemon.provider_check")
        .current_dir(&canon_daemon)
        .env("PYTHONPATH", &canon_daemon)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

    crate::hide_window(&mut cmd);

    let mut child = cmd
        .spawn()
        .map_err(|e| format!("Failed to launch provider checker: {e}"))?;

    if let Some(stdin) = child.stdin.as_mut() {
        stdin
            .write_all(payload.to_string().as_bytes())
            .map_err(|e| format!("Failed to send validation payload: {e}"))?;
    }
    let _ = child.stdin.take();

    let output = child
        .wait_with_output()
        .map_err(|e| format!("Failed to wait for provider checker: {e}"))?;

    let stdout = String::from_utf8_lossy(&output.stdout);
    let stderr = String::from_utf8_lossy(&output.stderr);

    if !output.status.success() && stdout.trim().is_empty() {
        return Ok(serde_json::json!({
            "ok": false,
            "code": "request_failed",
            "detail": stderr.trim().chars().take(300).collect::<String>(),
        }));
    }

    serde_json::from_str(&stdout)
        .map_err(|_| format!("Failed to parse provider validation result: {}", stderr.trim()))
}
